esp/platform/espprotocol.cpp

```diff
@@ -755,3 +755,92 @@ int CEspProtocol::countBindings(int port)
     else
         return apport->getBindingCount();
 }
+
+//Class to detect if connection is closed or reset by peer
+class ConnectionCloseDetector : implements IConnectionCloseDetector, public CInterface
+{
+private:
+    class CloseDetectionThread : public Thread
+    {
+    private:
+        ISocket* m_sock;
+        bool m_stop = false;
+        ConnectionCloseDetector* m_detector;
+    public:
+        CloseDetectionThread(ISocket* sock, ConnectionCloseDetector* detector) : m_sock(sock), m_detector(detector) {}
+        void stop()
+        {
+            m_stop = true;
+            join();
+        }
+        virtual int run()
+        {
+            while (!m_stop)
+            {
+                int ret = m_sock->wait_read(100);
+                if (m_stop)
+                    break;
+                if (ret < 0)
+                {
+                    m_detector->notifyClose();
+                    break;
+                }
+                else if (ret > 0)
+                {
+                    if (m_sock->avail_read() == 0)
+                    {
+                        m_detector->notifyClose();
+                        break;
+                    }
+                    else
+                        Sleep(10); // avoid tight loop in case there're still outstanding data on the wire.
+                }
+            }
+            return 0;
+        }
+    };
```

Comment:
The idea that poll() returns ok and then the input queue is empty *suggests strongly* the socket was closed at the other end.  Probably works great in all our cases.
I'm not sure it would behave any differently, but do also wonder about using 
recv(MSG_DONTWAIT|MSG_PEEK)
instead ?
How long would we wait ?  Should keepalive be enabled so that socket remains ok until user explicitly calls close() ?

Reply:
Hi Mark, the use case of the function is for some long-running service methods (for example the WsWorkunit::WuResult method), where the user sometimes would stop the method from the browser, thus making it pointless for the service method to keep running. So the detector is simply a non-intrusive way to check if the connection is closed by the other end, and it's not responsible for keeping the connection alive. The detector waits till the connection is closed, or when the service method is completed (when the detector object is released to be exact), whichever comes first.

The jlib avail_read() function calls ioctl(sock, FIONREAD, &avail) underneath, and it works pretty well from tests. recv(MSG_DONTWAIT|MSG_PEEK) will likely work as good, but I think ioctl is slightly better here because it's less intrusive, in that we don't need to allocate a buffer to actually receive any data.

Reply:
Of course for tls avail_read() does use SSL_peek and SSL_pending, a solution that you and I worked out sometime ago...

Reply:
I was thinking of:
char data[2];
recv(fd, data, 1, MSG_DONTWAIT|MSG_PEEK)
But I agree its probably no better than ioctl(FIONREAD), I just wanted to mention this as an alternative that is maybe a little more portable :-)

Reply:
Does the parent (or any other) thread call wait_read() + read() on the same socket at the same time ?
If another thread is calling wait_read() + read() often, is there ever a chance of a race where the other thread reads all the data so this threads avail_read() returns 0 ?

Reply:
No this function will be called in the service method, after all the request data has all been read in.

Reply:
But its in a thread running in a loop ?

Reply:
No it's not, read is done and there's no more loop there...

Reply:
Isn't this a loop that happens in a thread ?
```
while(!stop)
    wait_read()
    if avail_read() == 0
       // socket must have been closed ...
```

Reply:
Yes this thread runs in a loop, but it's the only thread that does wait_read(), no other thread does wait_read() or read() at the same time...

Reply:
ok.